core/src/network/reducers/shutdown.rs

```diff
@@ -0,0 +1,31 @@
+use crate::{action::ActionWrapper, context::Context, network::state::NetworkState};
+use holochain_net_connection::{
+    net_connection::NetConnection,
+    protocol_wrapper::{ProtocolWrapper, TrackAppData},
+};
+use std::sync::Arc;
+pub fn reduce_shutdown(
```

Comment:
is lifecycle management for a stateful entity in our system part of the state of the entity (which is what reducers are for)?

i'm imagining if this was a database, then this would be like needing an `INSERT` query every time we open/close a connection to the database

could we use the topological evaluation that nicho added for lifecycle management?

i'm thinking something like this https://github.com/stuartsierra/component

Reply:
@thedavidmeister I would say yes only because the current architecture requires us to put the entity into an "is_shut_down" state rather than completely destroying it, because there may still be live references to it in other threads. This is the case both for Arc references in core and for references to the instance in the mock network.

Reply:
@maackle what does "is shut down" mean if there are "live references to it"?

could you explain a simple example of how that works?

Reply:
@thedavidmeister by "live references" I mainly mean `Arc`s, but also value-based references like in `MockSingleton::senders` where instances are indirectly referenced by a HashMap of Senders, the channels of which point to the associated NetWorker for the Instance (so even if the Instance is dropped, this can be a sort of dangling reference leading to a `SendError`).

So the problem is that we have the possibility of instances being in this invalid "shutdown" state where we really want them to be completely gone, but other things are still referring to them. That invalid state is still a state we need to track, as long as we can't be 100% sure that we've got no dangling references to the thing. 

I guess the only question is if we want to track that state in the reducer or not. Here's one way to argue for it: `Shutdown` is the state transition that specifies that all other future state transitions are invalid. (If we could ensure that there would be no future attempts to transition state, we wouldn't need to think of it this way.) I also might be interpreting the intent of the whole redux state system wrongly too, so if this ultimately seems wrong I'll change it.

Here are the main the things that can be "shutdown" but not fully terminated (dropped):

* **`Instance`**: is_shutdown means the action loop is broken and the thread is joined, so no more actions are processed. (But the `Instance` may still live on in other `Arc`s).
* **`NetworkState`**: is_shutdown means the `P2pNetwork` and `NetWorker` (part of the state) are stopped. However, `NetworkState` is part of the `State`, which is referred to by the `Context` (via `Arc`) which is referred to by the `Holochain` (via `Arc`), therefore we can't dispose of the `NetworkState` either.
* therefore **`Holochain`**'s is_shutdown means the contained `Instance` is shutdown (meaning no `call` can be done), and the network is stopped. However there will still be `Arc<Holochain>`s hanging around so we can't dispose of the `Holochain` or `Instance` entirely.

All of these end-of-life changes are stateful because until we can fully dispose of these entities, they may be acted upon by other threads and reducers.

Reply:
"we really want them to be completely gone, but other things are still referring to them"

this bit i don't totally understand, why are "we" not also the "other things"? what other things exist outside "our" control?

if we truly want something to be completely gone, then we should track everything referencing it and do a cleanup, working through dependencies right?

if something has a dependency on something that is "shutdown", how can it continue to operate? isn't the dependant thing now operating in an invalid state, regardless of how "live" it is?

Reply:
@thedavidmeister and I had a good chat about this which convinced me to remove the shutdown reducer and deal with shutdown outside of the action loop. We also talked about possible approaches for tracking component dependencies to have more certainty about how to shut things down, which we might consider down the road when things get more complex.

Reply:
Ah, I was a bit confused since I saw the shutdown reducer yesterday when glancing over this PR. Interesting, what convinced you to remove the shutdown reducer? That was the part that made sense to me.. ;) Without that you have to do bad things like writing the state from the outside.

The main reason (currently) to have any shutdown process is the network, right? Telling the network module that we unregister from the network is a state transition that should be facilitated by an action reduction.